DevDaveFrame/wavepool
# Customizable News Post Ads

### To Verify

1. Clone/pull, run `pipenv install`, `pipenv shell`, 
2. Run `python3 manage.py makemigrations` and `python3 manage.py migrate`
3. Run `python3 manage.py runserver`
2. Navigate to [http://127.0.0.1:8000/admin/](http://127.0.0.1:8000/admin/)
3. Sign into the Django CMS Admin
4. Navigate to the "Advertisements" admin page under "Advertising"
5. Click "Add Advertisement +"
6. Create a new ad using the form
7. Select any news stories from the checkboxes at the bottom of the form
8. Save the new ad
9. Navigate to [http://127.0.0.1:8000/](http://127.0.0.1:8000/)
10. Find the stories you selected either on the front page or in the archive
11. Select a story to see it displayed with the ad you created

### Commentary

Upon reflection, I made some weird choices with this one. If I had taken more time to plan, I think I would have just added the industry dive ad as an Advertisement, maybe added a system-use-only flag and set it as the default in the NewsPost ad field. As it is, I decided to write a function that would pass the default ad from the pre-existing ADS array. Then I wrote the ad field to be null by default and wrote a ternary statement to pass either the associated ad, or the default-grabber function. This caused problems with the difference in object shape between the default dictionary object and the custom ads, so I wrote an if-else block in the template to access the right property. It's not exactly production-ready, but it works.
